src/jobs/models/manager.ts

```diff
@@ -138,34 +161,129 @@ export class JobManager {
+            name: stageData.type,
+            xstate: persistenceSnapshot,
+            userMetadata: stageData.userMetadata,
+            status: StageOperationStatus.CREATED,
+          })),
+        },
+        data: {
+          ...jobBeforeUpdate.data,
+          stages: unifiedStagesInput,
+        },
+      },
+
+      include: {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        Stage: true, // Include the created stages in the return value
+      },
+    };
+    try {
+      const job = await this.prisma.job.update(queryBody);
+
+      return this.convertPrismaToJobResponse(job);
+    } catch (error) {
+      this.logger.error(`Failed adding stage to job with error: ${(error as Error).message}`);
+
+      throw error;
+    }
+  }
+
   /**
    * This method is used to get a job entity by its id from the database.
    * @param jobId unique identifier of the job.
    * @returns The job entity if found, otherwise null.
    */
-  public async getJobEntityById(jobId: string): Promise<Prisma.JobGetPayload<Record<string, never>> | null> {
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  public async getJobEntityById(jobId: string): Promise<Prisma.JobGetPayload<{ include: { Stage: true } }> | null> {
     const queryBody = {
       where: {
         id: jobId,
       },
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      include: { Stage: true },
     };
 
     const job = await this.prisma.job.findUnique(queryBody);
 
     return job;
   }
 
-  private convertPrismaToJobResponse(prismaObjects: Prisma.JobGetPayload<Record<string, never>>): JobModel {
-    const { data, creationTime, userMetadata, expirationTime, notifications, updateTime, ttl, xstate, ...rest } = prismaObjects;
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  private convertPrismaToJobResponse(prismaObjects: Prisma.JobGetPayload<{ include: { Stage: true } }>): JobModel {
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    const { data, creationTime, userMetadata, expirationTime, notifications, updateTime, ttl, xstate, Stage, ...rest } = prismaObjects;
 
     const transformedFields = {
       data: data as Record<string, never>,
       creationTime: creationTime.toISOString(),
-      userMetadata: userMetadata as Record<string, never>,
+      userMetadata: userMetadata as { [key: string]: unknown },
       expirationTime: expirationTime ? expirationTime.toISOString() : undefined,
       notifications: notifications as Record<string, never>,
       updateTime: updateTime.toISOString(),
       ttl: ttl ? ttl.toISOString() : undefined,
+      stages: Array.isArray(Stage)
+        ? Stage.map((stage) => {
+            // eslint-disable-next-line @typescript-eslint/naming-convention
+            const { xstate, name, job_id, ...rest } = stage;
+            return Object.assign({ type: name, jobId: job_id }, rest);
+          })
+        : [],
```

Comment:
i would write a combined logic with the one in stageManager if possible. it can be moved to helpers then exported.

Reply:
very nice idea! i did so